File: ui-automation/selenium-java/src/main/java/utils/DriverFactory.java
package utils;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.edge.EdgeDriver;
import org.openqa.selenium.edge.EdgeOptions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.Properties;
import java.io.FileInputStream;
import java.io.IOException;

/**
 * WebDriver factory class to manage browser instances
 */
public class DriverFactory {
    private static ThreadLocal<WebDriver> driverThreadLocal = new ThreadLocal<>();
    private static ThreadLocal<WebDriverWait> waitThreadLocal = new ThreadLocal<>();
    
    private static final int DEFAULT_TIMEOUT = 10;
    
    /**
     * Initialize WebDriver based on browser type
     */
    public static void initializeDriver(String browserName) {
        WebDriver driver = null;
        
        switch (browserName.toLowerCase()) {
            case "chrome":
                WebDriverManager.chromedriver().setup();
                ChromeOptions chromeOptions = new ChromeOptions();
                chromeOptions.addArguments("--disable-notifications");
                chromeOptions.addArguments("--disable-popup-blocking");
                driver = new ChromeDriver(chromeOptions);
                break;
                
            case "firefox":
                WebDriverManager.firefoxdriver().setup();
                FirefoxOptions firefoxOptions = new FirefoxOptions();
                firefoxOptions.addPreference("dom.webnotifications.enabled", false);
                driver = new FirefoxDriver(firefoxOptions);
                break;
                
            case "edge":
                WebDriverManager.edgedriver().setup();
                EdgeOptions edgeOptions = new EdgeOptions();
                edgeOptions.addArguments("--disable-notifications");
                driver = new EdgeDriver(edgeOptions);
                break;
                
            default:
                throw new IllegalArgumentException("Browser not supported: " + browserName);
        }
        
        driver.manage().window().maximize();
        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(DEFAULT_TIMEOUT));
        
        driverThreadLocal.set(driver);
        waitThreadLocal.set(new WebDriverWait(driver, Duration.ofSeconds(DEFAULT_TIMEOUT)));
    }
    
    /**
     * Get current WebDriver instance
     */
    public static WebDriver getDriver() {
        return driverThreadLocal.get();
    }
    
    /**
     * Get current WebDriverWait instance
     */
    public static WebDriverWait getWait() {
        return waitThreadLocal.get();
    }
    
    /**
     * Quit WebDriver and clean up thread local
     */
    public static void quitDriver() {
        WebDriver driver = driverThreadLocal.get();
        if (driver != null) {
            driver.quit();
            driverThreadLocal.remove();
            waitThreadLocal.remove();
        }
    }
}

File: ui-automation/selenium-javascript/utils/DriverFactory.js
const { Builder, By, until } = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');
const firefox = require('selenium-webdriver/firefox');
const edge = require('selenium-webdriver/edge');

class DriverFactory {
    constructor() {
        this.driver = null;
    }

    /**
     * Initialize WebDriver based on browser type
     * @param {string} browserName - Browser name (chrome, firefox, edge)
     */
    async initializeDriver(browserName = 'chrome') {
        const browser = browserName.toLowerCase();
        
        switch (browser) {
            case 'chrome':
                const chromeOptions = new chrome.Options();
                chromeOptions.addArguments('--disable-notifications');
                chromeOptions.addArguments('--disable-popup-blocking');
                chromeOptions.addArguments('--disable-dev-shm-usage');
                chromeOptions.addArguments('--no-sandbox');
                
                this.driver = await new Builder()
                    .forBrowser('chrome')
                    .setChromeOptions(chromeOptions)
                    .build();
                break;
                
            case 'firefox':
                const firefoxOptions = new firefox.Options();
                firefoxOptions.setPreference('dom.webnotifications.enabled', false);
                
                this.driver = await new Builder()
                    .forBrowser('firefox')
                    .setFirefoxOptions(firefoxOptions)
                    .build();
                break;
                
            case 'edge':
                const edgeOptions = new edge.Options();
                edgeOptions.addArguments('--disable-notifications');
                
                this.driver = await new Builder()
                    .forBrowser('MicrosoftEdge')
                    .setEdgeOptions(edgeOptions)
                    .build();
                break;
                
            default:
                throw new Error(`Browser not supported: ${browserName}`);
        }
        
        await this.driver.manage().window().maximize();
        await this.driver.manage().setTimeouts({ implicit: 10000 });
        
        return this.driver;
    }

    /**
     * Get current WebDriver instance
     */
    getDriver() {
        return this.driver;
    }

    /**
     * Quit WebDriver and clean up
     */
    async quitDriver() {
        if (this.driver) {
            await this.driver.quit();
            this.driver = null;
        }
    }

    /**
     * Navigate to URL
     */
    async navigateTo(url) {
        await this.driver.get(url);
    }

    /**
     * Get current URL
     */
    async getCurrentUrl() {
        return await this.driver.getCurrentUrl();
    }

    /**
     * Get page title
     */
    async getPageTitle() {
        return await this.driver.getTitle();
    }

    /**
     * Wait for element to be located
     */
    async waitForElement(locator, timeout = 10000) {
        return await this.driver.wait(until.elementLocated(locator), timeout);
    }

    /**
     * Wait for element to be visible
     */
    async waitForElementVisible(locator, timeout = 10000) {
        const element = await this.waitForElement(locator, timeout);
        return await this.driver.wait(until.elementIsVisible(element), timeout);
    }

    /**
     * Wait for element to be clickable
     */
    async waitForElementClickable(locator, timeout = 10000) {
        const element = await this.waitForElement(locator, timeout);
        return await this.driver.wait(until.elementIsEnabled(element), timeout);
    }
}

module.exports = DriverFactory;
